File: chrome/browser/share/link_to_text_bridge.cc
// Copyright 2020 The Chromium Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include <jni.h>

#include "components/shared_highlighting/core/common/disabled_sites.h"
#include "components/shared_highlighting/core/common/shared_highlighting_metrics.h"
#include "content/public/browser/web_contents.h"
#include "url/android/gurl_android.h"

// Must come after all headers that specialize FromJniType() / ToJniType().
#include "chrome/android/chrome_jni_headers/LinkToTextBridge_jni.h"

using shared_highlighting::LinkGenerationReadyStatus;
using shared_highlighting::LinkGenerationStatus;

namespace {

ukm::SourceId GetSourceId(
    const base::android::JavaRef<jobject>& j_web_contents) {
  content::WebContents* web_contents =
      content::WebContents::FromJavaWebContents(j_web_contents);
  return web_contents->GetPrimaryMainFrame()->GetPageUkmSourceId();
}
}  // namespace

// TODO(gayane): Update the name whenever
// |ShouldOfferLinkToText| updated to more descriptive
// name.
static bool JNI_LinkToTextBridge_ShouldOfferLinkToText(
    JNIEnv* env,
    const base::android::JavaRef<jobject>& j_url) {
  GURL url = url::GURLAndroid::ToNativeGURL(env, j_url);
  return shared_highlighting::ShouldOfferLinkToText(url);
}

static bool JNI_LinkToTextBridge_SupportsLinkGenerationInIframe(
    JNIEnv* env,
    const base::android::JavaRef<jobject>& j_url) {
  GURL url = url::GURLAndroid::ToNativeGURL(env, j_url);
  return shared_highlighting::SupportsLinkGenerationInIframe(url);
}

static void JNI_LinkToTextBridge_LogFailureMetrics(
    JNIEnv* env,
    const base::android::JavaRef<jobject>& j_web_contents,
    int32_t error) {
  shared_highlighting::LogRequestedFailureMetrics(
      GetSourceId(j_web_contents),
      static_cast<shared_highlighting::LinkGenerationError>(error));
}

static void JNI_LinkToTextBridge_LogSuccessMetrics(
    JNIEnv* env,
    const base::android::JavaRef<jobject>& j_web_contents) {
  shared_highlighting::LogRequestedSuccessMetrics(GetSourceId(j_web_contents));
}

static void JNI_LinkToTextBridge_LogLinkRequestedBeforeStatus(
    JNIEnv* env,
    int32_t status,
    int32_t ready_status) {
  shared_highlighting::LogLinkRequestedBeforeStatus(
      static_cast<LinkGenerationStatus>(status),
      static_cast<LinkGenerationReadyStatus>(ready_status));
}

static void JNI_LinkToTextBridge_LogLinkToTextReshareStatus(JNIEnv* env,
                                                            int32_t status) {
  shared_highlighting::LogLinkToTextReshareStatus(
      static_cast<shared_highlighting::LinkToTextReshareStatus>(status));
}

DEFINE_JNI(LinkToTextBridge)

File: chrome/browser/share/android/java/src/org/chromium/chrome/browser/share/link_to_text/LinkToTextBridge.java
// Copyright 2020 The Chromium Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package org.chromium.chrome.browser.share.link_to_text;

import org.jni_zero.NativeMethods;

import org.chromium.content_public.browser.WebContents;
import org.chromium.url.GURL;

/** A Java API for connecting to shared_highlighting component. */
public class LinkToTextBridge {
    // TODO(gayane): Update the name whenever |shared_highlighting::ShouldOfferLinkToText| updated
    // to more descriptive name.
    public static boolean shouldOfferLinkToText(GURL url) {
        return LinkToTextBridgeJni.get().shouldOfferLinkToText(url);
    }

    public static boolean supportsLinkGenerationInIframe(GURL url) {
        return LinkToTextBridgeJni.get().supportsLinkGenerationInIframe(url);
    }

    public static void logFailureMetrics(WebContents webContents, @LinkGenerationError int error) {
        LinkToTextBridgeJni.get().logFailureMetrics(webContents, error);
    }

    public static void logSuccessMetrics(WebContents webContents) {
        LinkToTextBridgeJni.get().logSuccessMetrics(webContents);
    }

    public static void logLinkRequestedBeforeStatus(
            @LinkGenerationStatus int status, @LinkGenerationReadyStatus int readyStatus) {
        LinkToTextBridgeJni.get().logLinkRequestedBeforeStatus(status, readyStatus);
    }

    public static void logLinkToTextReshareStatus(@LinkToTextReshareStatus int status) {
        LinkToTextBridgeJni.get().logLinkToTextReshareStatus(status);
    }

    @NativeMethods
    interface Natives {
        boolean shouldOfferLinkToText(GURL url);

        boolean supportsLinkGenerationInIframe(GURL url);

        void logFailureMetrics(WebContents webContents, @LinkGenerationError int error);

        void logSuccessMetrics(WebContents webContents);

        void logLinkRequestedBeforeStatus(
                @LinkGenerationStatus int status, @LinkGenerationReadyStatus int readyStatus);

        void logLinkToTextReshareStatus(@LinkToTextReshareStatus int status);
    }
}
